lib/Sema/CSSimplify.cpp

```diff
@@ -6565,35 +6499,50 @@ bool ConstraintSystem::resolveClosure(TypeVariableType *typeVar,
   auto *closureLocator = typeVar->getImpl().getLocator();
   auto *closure = cast<ClosureExpr>(closureLocator->getAnchor());
 
-  auto *closureType = getClosureType(closure);
-
-  auto *paramList = closure->getParameters();
-  for (unsigned i = 0, n = paramList->size(); i != n; ++i) {
-    const auto &param = closureType->getParams()[i];
+  auto *inferredClosureType = getClosureType(closure);
 
-    Type internalType;
+  auto isContextuallyInOut = [&contextualType](unsigned index) -> bool {
+    if (!contextualType)
+      return false;
 
-    if (paramList->get(i)->getTypeRepr()) {
-      // Internal type is the type used in the body of the closure,
-      // so "external" type translates to it as follows:
-      //  - `Int...` -> `[Int]`,
-      //  - `inout Int` -> `@lvalue Int`.
-      internalType = param.getParameterType();
-    } else {
-      auto *paramLoc =
-          getConstraintLocator(closure, LocatorPathElt::TupleElement(i));
+    if (auto *fnType = contextualType->getAs<FunctionType>()) {
+      if (fnType->getNumParams() > index)
+        return fnType->getParams()[index].isInOut();
+    }
 
-      internalType = createTypeVariable(paramLoc, TVO_CanBindToLValue |
-                                                      TVO_CanBindToNoEscape);
+    return false;
+  };
 
-      auto externalType = param.getOldType();
-      addConstraint(ConstraintKind::BindParam, externalType, internalType,
-                    paramLoc);
+  auto *paramList = closure->getParameters();
+  llvm::SmallVector<AnyFunctionType::Param, 4> parameters;
+  for (unsigned i = 0, n = paramList->size(); i != n; ++i) {
+    auto *PD = paramList->get(i);
+    const auto &param = inferredClosureType->getParams()[i];
+
+    // If there is no explicit "external" type for a given parameter
+    // and it's supposed to be `inout` contextually - let's adjust
+    // specifier of the declaration as well as closure function type
+    // to account of contextual information.
+    if (!PD->getTypeRepr() && !param.isInOut() && isContextuallyInOut(i)) {
+      auto flags = param.getParameterFlags();
+      parameters.push_back(param.withFlags(flags.withInOut(true)));
+      ModifiedSpecifiers.push_back(std::make_pair(PD, PD->getSpecifier()));
+      PD->setSpecifier(ParamSpecifier::InOut);
+    } else {
+      parameters.push_back(std::move(param));
     }
 
-    setType(paramList->get(i), internalType);
+    // Internal type is the type used in the body of the closure,
+    // so "external" type translates to it as follows:
+    //  - `Int...` -> `[Int]`,
+    //  - `inout Int` -> `@lvalue Int`.
+    setType(PD, param.getParameterType());
```

Comment:
`getParameterType()` will never give you `InOutType` or `LValueType`. This comment feels weird.